File: Front/TS/tests/unit/SUTs/CreateBetSut.ts
import { expect } from 'vitest';
import { Bet } from '../../../src/domain/bets/Bet';
import { CreateBetHandler, CreateBetResponse, ICreateBetRequest } from '../../../src/domain/features/CreateBetHandler';
import { FakeBetRepository } from '../implems/FakeBetRepository';
import { FakeCreateBetPresenter } from '../implems/FakeCreateBetPresenter';
import { StubDateTimeProvider } from '../implems/StubDateTimeProvider';
import { StubIdGenerator } from '../implems/StubIdGenerator';
export class CreateBetSut {
    
    private request: ICreateBetRequest;
    private betRepository: FakeBetRepository = new FakeBetRepository();
    private presenter = new FakeCreateBetPresenter;
    private idGenerator = new StubIdGenerator("id");
    private dateTimeProvider: StubDateTimeProvider;
    
    WithRequest(request: ICreateBetRequest) {
        this.request = request
        return this;
    }
    
    WithCurrentDate(currentDate: Date) {
        this.dateTimeProvider = new StubDateTimeProvider(currentDate)
        return this;
    }

    async WhenExecuteHandler() : Promise<CreateBetSut> {
        const handler = new CreateBetHandler(this.betRepository, this.presenter, this.idGenerator, this.dateTimeProvider);
        await handler.Handle(this.request);
        return this;
    }
    
    BetShouldBeCreated() {
        expect(this.betRepository.Bets).toContainEqual(new Bet(this.idGenerator.Generate(), 
                                                                this.request.Description, 
                                                                this.request.EndDate,
                                                                this.request.Chips,
                                                                this.request.Friends))
        expect(this.presenter.Bet).toEqual(new CreateBetResponse(this.idGenerator.Generate(), 
                                                                this.request.Description, 
                                                                this.request.EndDate,
                                                                this.request.Chips,
                                                                this.request.Friends))
    }
    
    BetIsNotCreated(error: string) {
        expect(this.betRepository.Bets).toEqual([]);
        expect(this.presenter.Bet).toBeUndefined();
        expect(this.presenter.Error).toBe(error);
    }
    
}

File: Front/.Net Maui/BetFriends.Domain.UnitTests/SUTs/CreateBetSut.cs
using BetFriends.Domain.Features.CreateBet;
using BetFriends.Domain.UnitTests.Implems;

namespace BetFriends.Domain.UnitTests.SUTs;

internal class CreateBetSut
{
    private StubDateTimeProvider dateTimeProvider = default!;
    private readonly MockBetRepository betRepository = new(default!);
    private CreateBetRequest request = default!;
    private readonly StubIdGenerator idGenerator = new StubIdGenerator("id");
    private readonly StubCreateBetPresenter presenter = new StubCreateBetPresenter();
    internal CreateBetSut WithDateProvider(StubDateTimeProvider stubDateTimeProvider)
    {
        dateTimeProvider = stubDateTimeProvider;
        return this;
    }

    internal CreateBetSut WithRequest(CreateBetRequest createBetRequest)
    {
        request = createBetRequest;
        return this;
    }

    internal async Task<CreateBetSut> WhenExecuteCommand()
    {
        var handler = new CreateBetRequestHandler(betRepository, dateTimeProvider, presenter, idGenerator);
        await handler.Handle(request, default!);
        return this;
    }

    internal void ShouldCreateBet()
    {
        Assert.Equal(new CreateBetResponse(idGenerator.Id, request.Description, request.EndDate, request.Chips, request.Friends), presenter.Bet);
        Assert.Contains(betRepository.Bets, x => x == new Bet(idGenerator.Id, request.Description, request.EndDate, request.Chips, request.Friends));
    }

    internal void ShouldNotCreateBet()
    {
        Assert.Empty(betRepository.Bets);
        Assert.Null(presenter.Bet);
    }
}
